### ISSUE TITLE
  How to configure iwyu to analyze header files in recursive way?

### ISSUE THREAD

**Issue by qzxuhui** [include-what-you-use/include-what-you-use#1204]
I want to configure iwyu to analyze the head file in a recursive way. It seems only one layer head file is be analyzed.
I apologize if it is a stuip problem.
I have a test case to demonstrate this.
```
// A.h
#ifndef A_H
#define A_H
class A
{
};
#endif
```

```
// B.h
#ifndef B_H
#define B_H
#include "A.h"
class B
{
public:
    A a;
};
#endif
```

```
// C.h
#ifndef C_h
#define C_h
#include "B.h"    // I think do not need to include A.h, because A.h already be included by B.h
class C
{
    A a;
    B b;
};
#endif
```
```
// C.cpp
#include "C.h"
void printC(C c)
{
}
```
```
# compile_commands.json
[
    {
        "directory": ".",
        "command": "g++ -c C.cpp",
        "file": "C.cpp"
    }
]
```
I think in `C.h` do not need to include `A.h`, because `C.h` already include `B.h`, `A.h` already be include by `B.h`.
However, when I using iwyu, with version `include-what-you-use 0.9 based on clang version 5.0.1-2 (tags/RELEASE_501/final)`, with command `iwyu_tool -p .`, the result showns that
```
C.h should add these lines:
#include "A.h"  // for A

C.h should remove these lines:

The full include-list for C.h:
#include "A.h"  // for A
#include "B.h"  // for B
---

(C.cpp has correct #includes/fwd-decls)
```
I can not understand why, or do I do something wrong? Thanks for your time.

**Comment by danielh2942** [comment]
Hey if you're using cmake
The VERIFY_INTERFACE_HEADER_SETS allows you to check and verify every IWYU header p easily

(I wrote the patch in cmake and it works for me at my job anyway)

**Comment by kimgr** [comment]
In this case I think it's intentional -- `C.h` is using both `A` and `B` explicitly, so the respective headers should be included explicitly. You can read more about the rationale for IWYU here: https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/WhyIWYU.md.

In this particular case, I think it's a matter of protecting `C.h` from changes to `B` -- if you only included `B.h`, and `B` was changed not to derive from `A`, `C.h` would no longer compile.

**Comment by correaa** [comment]
@danielh2942 ,

I am doing this ` cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_INCLUDE_WHAT_YOU_USE=iwyu -DCMAKE_VERIFY_INTERFACE_HEADER_SETS=ON` but it doesn't seem to have any effect in checking the headers (included in the tests for example).

**Comment by danielh2942** [comment]
@correaa 

are you using `target_include_directories` when defining the interface for the library?

As if you're using `include_directories` I'm not sure if it will work (I haven't had to verify headers in a while)

**Comment by correaa** [comment]
yes, I am using `target_include_directories` (or indirectly as `target_link_libraries`).
